File: ts-core/src/signing.ts
import { createSign, randomBytes } from 'node:crypto';
import { SigningError } from './exceptions.js';

/**
 * Fields excluded from the signing string, per the Telebirr specification.
 */
const EXCLUDED_FIELDS = [
  'sign',
  'sign_type',
  'header',
  'refund_info',
  'openType',
  'raw_request',
  'biz_content',
];

/**
 * Signs a Telebirr API request body using SHA256withRSA-PSS.
 *
 * Builds a canonical signing string from the request fields, then signs it
 * with the provided PKCS8 PEM private key using a 32-byte salt.
 *
 * @param request - The full request object (including `biz_content`).
 * @param privateKeyPem - PKCS8 PEM-encoded RSA private key.
 * @returns Base64-encoded signature string to set as the `sign` field.
 *
 * @example
 * ```ts
 * const sign = signRequest(
 *   { nonce_str: 'ABC', method: 'payment.preorder', biz_content: { appid: '123' } },
 *   privateKeyPem,
 * );
 * ```
 */
export function signRequest(
  request: Record<string, unknown>,
  privateKeyPem: string,
): string {
  const signString = buildSignString(request);
  return sha256PssSign(signString, privateKeyPem);
}

/**
 * Builds the canonical key=value signing string for a Telebirr request.
 *
 * Flattens top-level fields and `biz_content` inner fields, excludes
 * non-signable fields, sorts keys lexicographically, and joins as
 * `key=value&key=value`.
 *
 * @param request - The full request object.
 * @returns Sorted `key=value` pair string.
 *
 * @example
 * ```ts
 * buildSignString({ b: '2', a: '1', biz_content: { z: '9', m: '5' } });
 * // "a=1&b=2&m=5&z=9"
 * ```
 */
export function buildSignString(request: Record<string, unknown>): string {
  const fieldMap: Record<string, string> = {};
  const fields: string[] = [];

  for (const key of Object.keys(request)) {
    if (EXCLUDED_FIELDS.includes(key)) continue;
    fields.push(key);
    fieldMap[key] = String(request[key]);
  }

  const bizContent = request.biz_content as
    | Record<string, unknown>
    | undefined;
  if (bizContent && typeof bizContent === 'object') {
    for (const key of Object.keys(bizContent)) {
      if (EXCLUDED_FIELDS.includes(key)) continue;
      fields.push(key);
      fieldMap[key] = String(bizContent[key]);
    }
  }

  fields.sort();
  return fields.map((k) => `${k}=${fieldMap[k]}`).join('&');
}

/**
 * Signs a UTF-8 string with SHA256withRSA-PSS (32-byte salt).
 *
 * @param data - The plaintext string to sign.
 * @param privateKeyPem - PKCS8 PEM-encoded RSA private key.
 * @returns Base64-encoded signature.
 * @throws {SigningError} If the signing operation fails.
 *
 * @example
 * ```ts
 * const sig = sha256PssSign('appid=123&amount=100', privateKeyPem);
 * ```
 */
export function sha256PssSign(data: string, privateKeyPem: string): string {
  try {
    const sign = createSign('SHA256');
    sign.update(data);
    sign.end();
    const signature = sign.sign(
      {
        key: privateKeyPem,
        padding: 1, // RSA_PKCS1_PSS_PADDING
        saltLength: 32,
      },
      'base64',
    );
    return signature;
  } catch (err) {
    throw new SigningError(
      `Signing failed: ${err instanceof Error ? err.message : String(err)}`,
    );
  }
}

/**
 * Generates a random 32-character uppercase hexadecimal nonce string.
 *
 * Uses `crypto.randomBytes(16)` to produce 16 random bytes, converts
 * them to hex, and takes the first 32 characters.
 *
 * @returns A 32-character uppercase alphanumeric string.
 *
 * @example
 * ```ts
 * generateNonceStr(); // "A1B2C3D4E5F6A7B8C9D0E1F2A3B4C5D6"
 * ```
 */
export function generateNonceStr(): string {
  return randomBytes(16)
    .toString('hex')
    .toUpperCase()
    .slice(0, 32);
}

File: dart_core/lib/src/signing.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:cryptography/cryptography.dart';

import 'exceptions.dart';

/// Fields excluded from the signing string computation.
///
/// These fields are either metadata (like `sign` and `sign_type`) or
/// nested structures (like `biz_content`) whose inner fields are flattened
/// into the signing string instead.
const _excludedFields = {
  'sign',
  'sign_type',
  'header',
  'refund_info',
  'openType',
  'raw_request',
  'biz_content',
};

/// Builds the canonical signing string from a request body map.
///
/// The signing string is constructed by:
/// 1. Collecting all top-level fields except those in [_excludedFields].
/// 2. Flattening `biz_content`'s inner fields into the same level,
///    also excluding any fields in [_excludedFields].
/// 3. Sorting all keys lexicographically (ASCII sort).
/// 4. Joining as `key=value` pairs separated by `&`.
///
/// [request] is the full request body map (including `biz_content`).
///
/// Returns the canonical signing string ready to be signed.
///
/// Example:
/// ```dart
/// final signString = buildSignString({
///   'method': 'payment.preorder',
///   'biz_content': {
///     'appid': '12345',
///     'total_amount': '100',
///   },
/// });
/// // Returns 'appid=12345&method=payment.preorder&total_amount=100'
/// ```
String buildSignString(Map<String, dynamic> request) {
  final fieldMap = <String, String>{};
  final fields = <String>[];

  for (final key in request.keys) {
    if (_excludedFields.contains(key)) continue;
    fields.add(key);
    fieldMap[key] = request[key].toString();
  }

  final bizContent = request['biz_content'];
  if (bizContent is Map<String, dynamic>) {
    for (final key in bizContent.keys) {
      if (_excludedFields.contains(key)) continue;
      fields.add(key);
      fieldMap[key] = bizContent[key].toString();
    }
  }

  fields.sort();
  return fields.map((k) => '$k=${fieldMap[k]}').join('&');
}

/// Signs a Telebirr API request body using SHA256withRSA-PSS.
///
/// This is the main entry point for request signing. It:
/// 1. Builds the canonical signing string from [request] using [buildSignString].
/// 2. Signs the string with the RSA private key extracted from [privateKeyPem].
///
/// [request] is the full request body map (including `biz_content`).
/// [privateKeyPem] is the PKCS#8 PEM-encoded RSA private key.
///
/// Returns the Base64-encoded signature string.
///
/// Throws [SigningException] if signing fails (e.g., invalid key or
/// malformed PEM).
///
/// Example:
/// ```dart
/// final signature = await signRequest(body, privateKeyPem);
/// body['sign'] = signature;
/// ```
Future<String> signRequest(
  Map<String, dynamic> request,
  String privateKeyPem,
) async {
  final signString = buildSignString(request);
  return await _sha256PssSign(signString, privateKeyPem);
}

/// Performs SHA256withRSA-PSS signing with a 32-byte salt length.
///
/// Parses the PEM key, extracts PKCS#8 DER components, constructs
/// an [RsaKeyPairData], and signs using [RsaPss] from the `cryptography`
/// package.
///
/// [data] is the UTF-8 string to sign.
/// [privateKeyPem] is the PKCS#8 PEM-encoded RSA private key.
///
/// Returns the Base64-encoded signature.
///
/// Throws [SigningException] on any parsing or signing error.
Future<String> _sha256PssSign(String data, String privateKeyPem) async {
  try {
    final keyBytes = _pemToKeyBytes(privateKeyPem);
    final keyComponents = _parsePkcs8Der(keyBytes);

    final algorithm = RsaPss(
      Sha256(),
      nonceLengthInBytes: 32,
    );

    final keyPair = RsaKeyPairData(
      n: _bigIntToBytes(keyComponents['n']!),
      e: _bigIntToBytes(keyComponents['e']!),
      d: _bigIntToBytes(keyComponents['d']!),
      p: _bigIntToBytes(keyComponents['p']!),
      q: _bigIntToBytes(keyComponents['q']!),
      dp: _bigIntToBytes(keyComponents['dp']!),
      dq: _bigIntToBytes(keyComponents['dq']!),
      qi: _bigIntToBytes(keyComponents['qInv']!),
    );

    final signature = await algorithm.sign(
      utf8.encode(data),
      keyPair: keyPair,
    );

    return base64.encode(signature.bytes);
  } on TelebirrException {
    rethrow;
  } catch (e) {
    throw SigningException('Signing failed: $e');
  }
}

/// Converts a [BigInt] to a big-endian byte list.
///
/// Returns `[0]` for [BigInt.zero]. No leading zero bytes are added.
List<int> _bigIntToBytes(BigInt number) {
  if (number == BigInt.zero) return [0];
  final hex = number.toRadixString(16);
  final paddedHex = hex.length.isOdd ? '0$hex' : hex;
  final bytes = <int>[];
  for (var i = 0; i < paddedHex.length; i += 2) {
    bytes.add(int.parse(paddedHex.substring(i, i + 2), radix: 16));
  }
  return bytes;
}

/// Strips PEM header/footer and decodes the Base64 content into raw bytes.
///
/// Supports both `BEGIN PRIVATE KEY` (PKCS#8) and
/// `BEGIN RSA PRIVATE KEY` (PKCS#1) PEM formats.
Uint8List _pemToKeyBytes(String pem) {
  final cleaned = pem
      .replaceAll('-----BEGIN PRIVATE KEY-----', '')
      .replaceAll('-----END PRIVATE KEY-----', '')
      .replaceAll('-----BEGIN RSA PRIVATE KEY-----', '')
      .replaceAll('-----END RSA PRIVATE KEY-----', '')
      .replaceAll('\n', '')
      .replaceAll('\r', '')
      .replaceAll(' ', '');
  return base64.decode(cleaned);
}

/// Parses a PKCS#8 DER-encoded RSA private key into its component big integers.
///
/// Extracts the standard RSA key components: `n`, `e`, `d`, `p`, `q`,
/// `dp`, `dq`, and `qInv`.
///
/// [der] is the raw DER byte array (from PKCS#8 wrapping).
///
/// Returns a map of component name to [BigInt] value.
///
/// Throws [SigningException] if the DER structure is malformed or
/// uses an unsupported version.
Map<String, BigInt> _parsePkcs8Der(Uint8List der) {
  var offset = 0;

  List<int> readElement() {
    if (offset >= der.length) {
      throw SigningException('Unexpected end of DER data');
    }

    final tag = der[offset++];
    int length;

    if (offset >= der.length) {
      throw SigningException('Unexpected end of DER length');
    }

    final firstByte = der[offset++];

    if (firstByte & 0x80 == 0) {
      length = firstByte;
    } else {
      final numLengthBytes = firstByte & 0x7F;
      if (numLengthBytes == 0) {
        throw SigningException('Indefinite length not supported');
      }
      length = 0;
      for (var i = 0; i < numLengthBytes; i++) {
        if (offset >= der.length) {
          throw SigningException('Unexpected end of DER length bytes');
        }
        length = (length << 8) | der[offset++];
      }
    }

    if (offset + length > der.length) {
      throw SigningException(
        'DER element extends beyond data: need $length bytes at offset $offset',
      );
    }

    final content = der.sublist(offset, offset + length);
    offset += length;
    return [tag, ...content];
  }

  BigInt readInteger(List<int> element) {
    if (element.isEmpty || element[0] != 0x02) {
      throw SigningException('Expected INTEGER tag (0x02), got 0x${element[0].toRadixString(16)}');
    }
    final intContent = element.sublist(1);

    var result = BigInt.zero;
    for (final byte in intContent) {
      result = (result << 8) | BigInt.from(byte);
    }
    return result;
  }

  List<int> readSequence() {
    final element = readElement();
    if (element.isEmpty || element[0] != 0x30) {
      throw SigningException('Expected SEQUENCE tag (0x30), got 0x${element[0].toRadixString(16)}');
    }
    return element.sublist(1);
  }

  void skipElement() {
    readElement();
  }

  try {
    // Outer SEQUENCE
    readSequence();
    offset = 0;

    // Version INTEGER
    final version = readInteger(readElement());
    if (version != BigInt.zero) {
      throw SigningException('Unsupported PKCS8 version: $version');
    }

    // AlgorithmIdentifier SEQUENCE
    final algIdData = readSequence();
    final savedAlgOffset = offset;
    offset = 0;

    // Skip OID
    readElement();
    offset = algIdData.length;

    // Check for NULL
    if (offset < algIdData.length) {
      skipElement();
    }

    offset = savedAlgOffset + algIdData.length;

    // OCTET STRING (contains the actual private key)
    final octetElement = readElement();
    if (octetElement.isEmpty || octetElement[0] != 0x04) {
      throw SigningException('Expected OCTET STRING tag (0x04)');
    }
    offset = 0;

    // Inner SEQUENCE (RSAPrivateKey)
    final innerSeq = readElement();
    if (innerSeq.isEmpty || innerSeq[0] != 0x30) {
      throw SigningException('Expected inner SEQUENCE');
    }
    offset = 0;

    // version
    final innerVersion = readInteger(readElement());
    if (innerVersion != BigInt.zero) {
      throw SigningException('Unsupported RSA version: $innerVersion');
    }

    // n
    final n = readInteger(readElement());
    // e
    final e = readInteger(readElement());
    // d
    final d = readInteger(readElement());
    // p
    final p = readInteger(readElement());
    // q
    final q = readInteger(readElement());
    // dp
    final dp = readInteger(readElement());
    // dq
    final dq = readInteger(readElement());
    // qInv
    final qInv = readInteger(readElement());

    return {
      'n': n,
      'e': e,
      'd': d,
      'p': p,
      'q': q,
      'dp': dp,
      'dq': dq,
      'qInv': qInv,
    };
  } on TelebirrException {
    rethrow;
  } catch (e) {
    if (e is TelebirrException) rethrow;
    throw SigningException('Failed to parse PKCS8 DER: $e');
  }
}
